src/ir/bits.h

```diff
@@ -128,35 +128,92 @@ struct DummyLocalInfoProvider {
 template<typename LocalInfoProvider = DummyLocalInfoProvider>
 Index getMaxBits(Expression* curr,
                  LocalInfoProvider* localInfoProvider = nullptr) {
-  if (auto* const_ = curr->dynCast<Const>()) {
+  if (auto* c = curr->dynCast<Const>()) {
     switch (curr->type.getBasic()) {
       case Type::i32:
-        return 32 - const_->value.countLeadingZeroes().geti32();
+        return 32 - c->value.countLeadingZeroes().geti32();
       case Type::i64:
-        return 64 - const_->value.countLeadingZeroes().geti64();
+        return 64 - c->value.countLeadingZeroes().geti64();
       default:
         WASM_UNREACHABLE("invalid type");
     }
   } else if (auto* binary = curr->dynCast<Binary>()) {
     switch (binary->op) {
       // 32-bit
-      case AddInt32:
-      case SubInt32:
-      case MulInt32:
-      case DivSInt32:
-      case DivUInt32:
-      case RemSInt32:
-      case RemUInt32:
       case RotLInt32:
       case RotRInt32:
+      case SubInt32:
+        return 32;
+      case AddInt32: {
+        auto maxBitsLeft = getMaxBits(binary->left, localInfoProvider);
+        auto maxBitsRight = getMaxBits(binary->right, localInfoProvider);
+        return std::min(Index(32), std::max(maxBitsLeft, maxBitsRight) + 1);
+      }
+      case MulInt32: {
+        auto maxBitsRight = getMaxBits(binary->right, localInfoProvider);
+        auto maxBitsLeft = getMaxBits(binary->left, localInfoProvider);
+        return std::min(Index(32), maxBitsLeft + maxBitsRight);
+      }
+      case DivSInt32: {
+        if (auto* c = binary->right->dynCast<Const>()) {
+          auto maxBitsLeft = getMaxBits(binary->left, localInfoProvider);
+          if (maxBitsLeft == 32) {
+            return 32;
+          }
+          if (maxBitsLeft == 0) {
+            return 0;
+          }
+          auto value = c->value.geti32();
+          auto maxBitsRight = 31 - Index(CountLeadingZeroes(value));
+          return std::max(Index(0), maxBitsLeft - maxBitsRight);
```

Comment:
The type here looks wrong. the maxBits are both unsigned, so the difference is unsigned too. So if it would be negative, it will be a very big positive number, and "win" in the max operation.

Aside from the type it also looks wrong. We know that 
```
left <= 2^maxBitsLeft
right <= 2^maxBitsRight
```
(from the definition of maxBits). but that doesn't imply
```
left / right <= 2^(maxBitsLeft - maxBitsRight)
```
The only inequality I can almost see how to prove is
```
left / right <= 2^maxBitsLeft
```
but even that isn't quite right, as `right` may be 0, even if `maxBitsRight > 0`. This is similar to the issue from before: maxBits is an upper bound.

Reply:
Hmm. Perhaps it's not so simple. Decide check how do this in souper and it's look [more complex](https://github.com/google/souper/blob/41da4101160b786caa069c30bcd29add3fb6fd32/lib/Infer/AbstractInterpreter.cpp#L135)

Reply:
So here uses `31 + getMinBits(left) - getMaxBits(rhs)`

Reply:
> but even that isn't quite right, as right may be 0, even if maxBitsRight > 0. This is similar to the issue from before: maxBits is an upper bound.

It's easily could be rejected due to current implementation do max bits div calcs only for `x / C(onst)`: https://github.com/WebAssembly/binaryen/pull/2879/files#diff-58b7cf13906e9b5a4ba6306aaea394aaR158. Right value always const in my scenario. I don't try calc bits for general case like `x / y`